<a href="https://colab.research.google.com/github/cesaralejon/training_testing/blob/main/Divisi%C3%B3n_entrenar_probar_Cesar_Olmedo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [3]:
filename = ('/content/drive/MyDrive/Coding Dojo/Week 5/abalone.data')
df = pd.read_csv(filename, delimiter=',', header=None)
df.head()

0      1      2      3       4       5       6      7   8
0  M  0.455  0.365  0.095  0.5140  0.2245  0.1010  0.150  15
1  M  0.350  0.265  0.090  0.2255  0.0995  0.0485  0.070   7
2  F  0.530  0.420  0.135  0.6770  0.2565  0.1415  0.210   9
3  M  0.440  0.365  0.125  0.5160  0.2155  0.1140  0.155  10
4  I  0.330  0.255  0.080  0.2050  0.0895  0.0395  0.055   7

In [4]:
df.columns = ['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight','Shell weight','Rings']

In [5]:
df.head()

Sex  Length  Diameter  ...  Viscera weight  Shell weight  Rings
0   M   0.455     0.365  ...          0.1010         0.150     15
1   M   0.350     0.265  ...          0.0485         0.070      7
2   F   0.530     0.420  ...          0.1415         0.210      9
3   M   0.440     0.365  ...          0.1140         0.155     10
4   I   0.330     0.255  ...          0.0395         0.055      7

[5 rows x 9 columns]

In [14]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

In [11]:
x= df[['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight','Shell weight']]
y= df['Rings']

## Regresion lineal

Sin una base de prueba

In [12]:
reg = LinearRegression()
reg.fit(x, y)

LinearRegression()

In [13]:
reg.score(x,y)

0.5276299399919839

Con base de prueba

In [16]:
x_train, x_test, y_train, y_test = train_test_split(x,y, random_state = 2)

In [17]:
reg_bp = LinearRegression()
reg_bp.fit(x_train, y_train)

LinearRegression()

In [18]:
reg_bp.score(x_train, y_train)

0.5278063073537527

In [20]:
reg_bp.score(x_test, y_test) #probar el r2

0.5108074629638203

## KNN

sin base de prueba

In [23]:
scalar = StandardScaler()
scalar.fit(x)


StandardScaler()

In [24]:
x_scalar = scalar.transform(x)

In [25]:
knn = KNeighborsRegressor()

In [26]:
knn.fit(x_scalar, y)

KNeighborsRegressor()

In [28]:
knn.score(x_scalar, y) #probar el r2

0.6576046052340014

con base de prueba

In [29]:
x_train, x_test, y_train, y_test = train_test_split(x,y, random_state = 2)

In [30]:
scalar= StandardScaler()
scalar.fit(x_train)

StandardScaler()

In [31]:
x_train = scalar.transform(x_train)
x_test = scalar.transform(x_test)

In [33]:
knn_bp = KNeighborsRegressor()
knn_bp.fit(x_train, y_train)

KNeighborsRegressor()

In [34]:
knn_bp.score(x_train, y_train)

0.6537694311063846

Una vez que hayas completado la tarea, responde las siguientes preguntas:

* ¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando no usaste la división entrenar probar?

En regresión lineal el score de r2 = 0.5276299399919839

En KNN el score de r2 = 0.6576046052340014

El mejor modelo es KNN

* ¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando usaste la división entrenar probar?

En regresión lineal el training de r2 = 0.5278063073537527
En regresión lineal el score de r2 = 0.5108074629638203

En KNN el training de r2 = 0.6537694311063846

En términos generales es KNN, pero para poder determinar es necesarios calcular el mean squared error


* ¿Hubo alguna ventaja en la regresión lineal en términos de la cantidad de código que tenías que escribir?

Considero que la cantidad de código no tiene diferencia significativa apesar de que que en knn tenemos que escalar los datos


* ¿Hay alguna forma de mostrarle a alguien cuál de los dos modelos fue más efectivo?

Al comparar las métricas podemos indicar cual es más efectivo con respecto a la base de prueba.


* ¿Crees que podrías haber mejorado KNN para mejorar la eficacia del modelo?
Se pueden modificar el número de vecinos cercanos
